Add documentation for emacs configuration (#5)
bbannier/zeek-language-server: emacs/README.md

> # zeek-language-server

Reviewer: ```suggestion
# Emacs setup for zeek-language-server

```

Let's also link this from the top-level README.

Since this directory contains only a README file for now, I am wondering whether just adding a top-level `README.emacs` would be sufficient.

Author: I'd be alright with just moving it up.

> This directory contains the configuration files needed to tie into `zeek-language-server` from emacs. These instructions use `lsp-mode`. There's the possibility it would also work with `elgot` (an `lsp-mode` replacement), but that exercise is left to the reader. `lsp-mode` comes with an astonishing amount of options for configuring the look and operation of the LSP features in emacs. This guide only presents the most basic configuration.

Reviewer: ```suggestion
This file documents needed configuration to interface emacs with `zeek-language-server`. These instructions use `lsp-mode`. There's the possibility it would also work with `elgot` (an `lsp-mode` replacement), but that exercise is left to the reader. `lsp-mode` comes with an astonishing amount of options for configuring the look and operation of the LSP features in emacs. This guide only presents the most basic configuration.

```

This guide assumes that you have installed `zeek-language-server` and that it is in your `PATH`.

## Prerequisites

There are some emacs packages that are required for this to work. You can install all of these from your favorite emacs package repository using `M-x package-install`. The required packages are:

```
use-package
lsp-mode
lsp-ui

> yasnippet

Reviewer: I am not sure how familiar emacs users are with packages, but when I dropped below basic config into my system it failed (packages were not already present, and not installed automatically). When I installed them with `M-x package-install` things progressed.

Author: Reworded the section a little bit to mention that they should all be installed before adding the snippet.

```

It also requires the Zeek major mode from https://github.com/zeek/emacs-zeek-mode. Either copy the file into a directory in your emacs `load-path` or load the file manually like below in your `.emacs`:

```
(load-file "/path/to/zeek-mode")

> ```

Reviewer: Is this actually a hard requirement, or just nice to have? Since this file isn't in the upstream package repos this seems to introduce some complications.

If we include it we need to make sure that formatting from `emacs-zeek-mode` does not interfere with formatting from the LSP server. Do you know whether this is disabled in `lsp-mode` by default (which would be fine)?

Author: It is, because we need zeek-mode for the hook to lsp-mode to work.

Reviewer: > As for the formatting, I'm not sure. I'd have to play with it.

Looks like format-on-save is by default also disabled in `emacs-zeek-mode`, so the default config has no conflict. Unless behavior of one of the involved plugins change, there's probably no issue right now unless users go out of their way to mess things up.

> My only concern with doing that is that it's different from how most major modes handle things.

I am also wondering how that would impact users' ability to configure LSP integration. Like you write, this plugin comes with **a lot** of options, and users will likely want to tweak things.

A basic configuration for all of the necessary packages is below. You should be able to drop this directly into your `.emacs`, preferably near the top.

```
(require 'package)
(add-to-list 'package-archives
	     '("melpa-stable" . "https://stable.melpa.org/packages/") t)

(package-initialize)

(eval-when-compile
  (require 'use-package))

(use-package lsp-mode
	     :commands lsp)
(use-package lsp-ui)
(use-package yasnippet)
(use-package zeek-mode)
```

## Setup

> 1. Add the path to `zeek-config` (likely in your Zeek build directory) to your environment's `PATH` variable.

Reviewer: I am not sure how common it is to not install Zeek from packages. Maybe instead
```suggestion
1. If needed, add the path to `zeek-config`to your environment's `PATH` variable.

```

Author: Fixed.

2. Add a hook to `lsp-mode` for zeek files

```
(add-hook 'zeek-mode-hook 'lsp)
```

3. Register `zeek-language-server` with `lsp-mode`

```
(with-eval-after-load 'lsp-mode
  (add-to-list 'lsp-language-id-configuration
    '(zeek-mode . "zeek"))

  (lsp-register-client
    (make-lsp-client :new-connection (lsp-stdio-connection "zeek-language-server")
                     :activation-fn (lsp-activate-on "zeek")
                     :server-id 'zeek)))
```

> 4. Open a zeek script file in emacs. At this point you should notice two things about your emacs window. First, in the modeline it should list that it is using the zeek major-mode and that it connected to the LSP server. These are shown something like `(zeek LSP[zeek:pid])`. Second, you should have line at the top of the window that gives information about where in the Zeek file you are.

Reviewer: I still seem to be missing a number of features I'd expect with this config. 

Works:
- when writing function calls I get a signature helper near the bottom, but I don't seem to be getting completion

Missing:
- no completions
- diagnostics missing: syntax errors in unsaved buffers not reported (should show diagnostics from tree-sitter parser); syntax errors in saved buffers not reported (should show diagnostics from `zeek --parse-only`)

This is with Homebrew emacs-27.2.

Author: Completion depends on `company-mode`, which I didn't include in the guide (mostly because I haven't set it up myself yet). Diagnostics might be available in another buffer. Hit `ctrl-x right-arrow` to scroll between them.

Reviewer: Just installing the `company` package got me completions, I think we should add it to the docs.

I couldn't find any diagnostics buffer, but `flymake-mode` seems to report something (`Flymake[0 0 1]` in modeline).

> 

Reviewer: Could you add a pointer to `lsp-mode` (?) docs showing commands to get more information.

At least I have no understanding how to get access to other capabilities exposed by the server (hover, goto declaration/definition/implementation, list symbols, ...).

Author: Added to the header.

> ## TODO
> 
> - Automatically configure the path to `zeek-config`. This would be especially useful when switching between multiple clones of the Zeek repository.

Reviewer: Suggest to drop this as I suspect this is not a widespread use case.
```suggestion
```

Author: Dropped.
